<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [347]:
import pandas as pd

In [348]:
# Install this libraries if they are not installed. If installed, you dont have to run them.
# !pip install psycopg2-binary
# !pip install sqlalchemy

### Setting up parameters for connection to database

In [349]:
param = {}

#open config.text that contains credentials to connect to database
with open ('config.txt') as file:
    
    #read each line of the text file
    for line in file:
        
        #split the line by '=' and save into variables key, value
        key, value = line.split('=')
        
        #remove any leading or trailing white spaces
        key, value = key.strip(), value.strip()
        
        #add key-value pairs to the dictionary
        param[key] = value


In [350]:
#connection string with credentials taken from param dictionary
connection = f'postgresql+psycopg2://{param["USER"]}:{param["PASSWORD"]}@{param["HOST"]}:{param["PORT"]}/{param["DBNAME"]}'

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [351]:
# текст запроса
query_3_1 = '''
select 
    count(distinct id)
from vacancies
'''

In [352]:
# результат запроса
answer_3_1 = pd.read_sql_query(query_3_1, connection)
print(f"3.1 Количество уникальных доступных вакансий: {answer_3_1['count'][0]}")

3.1 Количество уникальных доступных вакансий: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [353]:
# текст запроса
query_3_2 = '''
select 
    count(distinct id)
from employers
'''

In [354]:
# результат запроса
answer_3_2 = pd.read_sql_query(query_3_2, connection)
print(f"3.2 Количество уникальных работодателей: {answer_3_2['count'][0]}")

3.2 Количество уникальных работодателей: 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [355]:
# текст запроса
query_3_3 = '''
select 
    count(distinct id)
from areas
'''

In [356]:
# результат запроса
answer_3_3 = pd.read_sql_query(query_3_3, connection)
print(f"3.3 Количество уникальных географических зон с доступными вакансиями: {answer_3_3['count'][0]}")

3.3 Количество уникальных географических зон с доступными вакансиями: 1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [357]:
# текст запроса
query_3_4 = '''
select
    count(distinct industry_id)
from employers_industries
'''

In [358]:
# результат запроса
answer_3_4 = pd.read_sql_query(query_3_4, connection)
print(f"3.4 Количество уникальных отраслей в таблице отраслей работодателей: {answer_3_4['count'][0]}")

3.4 Количество уникальных отраслей в таблице отраслей работодателей: 294


***

### выводы по предварительному анализу данных:
### Общий обзор (Раздел 3)

- **Разнообразие возможностей (3.1)**: С 49 197 уникальных вакансий, рынок труда выглядит живым и разнообразным.
- **Разнообразие работодателей (3.2)**: Значительное количество уникальных работодателей (23 501) указывает на здоровую конкуренцию за таланты.
- **Географическое распределение (3.3)**: 1 362 уникальных района показывают, что возможности для работы не ограничены лишь несколькими регионами.
- **Диверсификация отраслей (3.4)**: С 294 уникальными отраслями, рынок не доминируется какой-либо одной сферой, предлагая широкий спектр карьерных путей.

### Заключения:

1. **Динамичный рынок**: Огромное количество вакансий и работодателей говорит о динамичном, конкурентоспособном рынке труда.
2. **Географическая гибкость**: Широкое географическое распределение может быть преимуществом, если вы открыты к переезду или ищете удаленную работу.
3. **Межотраслевые возможности**: Разнообразие отраслей означает, что вы потенциально можете переключиться или исследовать разные сферы.



### English:
### General Overview (Section 3)

- **Variety of Opportunities (3.1)**: With 49,197 unique vacancies, the job market appears robust and diverse.
- **Employer Diversity (3.2)**: A significant number of unique employers (23,501) suggests a healthy competition for talent.
- **Geographical Spread (3.3)**: 1,362 unique areas indicate that job opportunities are not confined to just a few regions.
- **Industry Diversification (3.4)**: With 294 unique industries, the market isn't dominated by any single sector, offering a wide range of career paths.

### Conclusions:

1. **Dynamic Market**: The sheer number of vacancies and employers suggests a dynamic, competitive job market.
2. **Geographical Flexibility**: The wide geographic spread could be advantageous if you're open to relocating or looking for remote work.
3. **Cross-Industry Opportunities**: The diversity in industries means you could potentially pivot or explore different fields.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [359]:
# текст запроса
query_4_1 = '''
select 
    a.name area_name,
    count(v.id) cnt
from areas a
    left join vacancies v on v.area_id = a.id
group by a.name
order by count(v.id) desc
'''

In [360]:
# результат запроса
answer_4_1 = pd.read_sql_query(query_4_1, connection)
print("4.1 Количество доступных вакансий в каждом регионе:")
answer_4_1.head(20)

4.1 Количество доступных вакансий в каждом регионе:


area_name   cnt
0            Москва  5333
1   Санкт-Петербург  2851
2             Минск  2112
3       Новосибирск  2006
4            Алматы  1892
5      Екатеринбург  1698
6   Нижний Новгород  1670
7            Казань  1415
8         Краснодар  1301
9            Самара  1144
10   Ростов-на-Дону  1131
11          Воронеж  1063
12          Ташкент   851
13       Красноярск   847
14       Нур-Султан   840
15        Челябинск   786
16            Пермь   771
17              Уфа   767
18            Томск   692
19             Омск   617

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [361]:
# текст запроса
query_4_2 = '''
select
    count(*) as cnt
from vacancies v
where
    v.salary_from is not null or v.salary_to is not null
'''

In [362]:
# результат запроса
answer_4_2 = pd.read_sql_query(query_4_2, connection)
print(f"4.2 Количество вакансий с информацией о зарплате, либо начальной, либо максимальной, либо обеих: {answer_4_2['cnt'][0]}")

4.2 Количество вакансий с информацией о зарплате, либо начальной, либо максимальной, либо обеих: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [363]:
# текст запроса
query_4_3 = '''
select
    round(avg(v.salary_from), 0) as avg_starting_salary,
    round(avg(v.salary_to), 0) as avg_top_salary
from vacancies v
where v.salary_from is not null or v.salary_to is not null
'''

In [364]:
# результат запроса
answer_4_3 = pd.read_sql_query(query_4_3, connection)
print('4.3 Средние начальная и верхняя зарплаты для доступных вакансий:')
answer_4_3

4.3 Средние начальная и верхняя зарплаты для доступных вакансий:


avg_starting_salary  avg_top_salary
0              71065.0        110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [365]:
# текст запроса
query_4_4 = '''
select
    v.schedule as work_schedule,
    v.employment as employment_type,
    count(v.employment) cnt

from vacancies v
group by v.schedule, v.employment
order by cnt desc
'''

In [366]:
# результат запроса
answer_4_4 = pd.read_sql_query(query_4_4, connection)
print('4.4 Количество доступных вакансий для каждой комбинации графика работы и типа занятости:')
answer_4_4

4.4 Количество доступных вакансий для каждой комбинации графика работы и типа занятости:


work_schedule      employment_type    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [367]:
# текст запроса
query_4_5 = '''
 select
    v.experience,
    count(*) cnt

from vacancies v
group by v.experience
order by cnt 
'''

In [368]:
# результат запроса
answer_4_5 = pd.read_sql_query(query_4_5, connection)
print('4.5 Количество доступных вакансий, сгруппированных по требуемому уровню опыта:')
answer_4_5

4.5 Количество доступных вакансий, сгруппированных по требуемому уровню опыта:


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

### выводы по детальному анализу вакансий
### Географическое Распределение (Таблица 4.1)

- **Крупные Города**: Большинство рабочих мест сосредоточено в Москве и Санкт-Петербурге.
- **Развивающиеся Рынки**: Меньшие города, такие как Новосибирск и Алматы, набирают обороты.

### Информация о Зарплатах (Таблица 4.3)

- **Средние Зарплаты**: Средняя начальная зарплата составляет около 71,065, а средняя верхняя зарплата — около 110,537. Это дает представление о том, какой диапазон зарплат ожидать.

### График Работы и Тип Занятости (Таблица 4.4)

- **Полная Занятость**: Большинство рабочих мест предполагают полную занятость и работу на полный рабочий день.
- **Удаленная Работа**: Количество удаленных рабочих мест также значительно, особенно для полной занятости.

### Уровень Опыта (Таблица 4.5)

- **Для Новичков**: Многие рабочие места открыты для кандидатов с опытом работы от 1 до 3 лет, затем идут те, которые требуют от 3 до 6 лет опыта.
- **Дружелюбны к Новичкам**: Достаточно большое количество рабочих мест также принимают кандидатов без опыта работы.

### Общие Выводы:

1. **Местоположение Имеет Значение**: Если вы ищете больше возможностей для работы, крупные города предлагают больше вариантов.
2. **Ожидания по Зарплате**: Таблица средних зарплат может служить полезным ориентиром для вас.
3. **Гибкость в Работе**: Достаточно большое количество рабочих мест предлагает удаленную работу и гибкий график, что может быть полезно, если вы хотите лучше совмещать работу и личную жизнь.
4. **Дружелюбны к Новичкам**: Рынок приветлив к людям с меньшим опытом, что отлично для тех, кто переходит в другую сферу или только начинает.


### English:
### Geographical Distribution (Table 4.1)

- **Major Cities**: The majority of jobs are concentrated in Moscow and St. Petersburg.
- **Emerging Markets**: Smaller cities like Novosibirsk and Almaty are catching up.

### Salary Insights (Table 4.3)

- **Average Salaries**: The average starting salary is about 71,065 and the average top salary is around 110,537. This gives you an idea of the salary range to expect.

### Work Schedule and Employment Type (Table 4.4)

- **Full-Time Jobs**: Most jobs are full-time and require a full-day commitment.
- **Remote Work**: The number of remote jobs is also significant, especially for full-time positions.

### Experience Level (Table 4.5)

- **Entry-Level**: Many jobs are open to candidates with 1-3 years of experience, followed by those requiring 3-6 years.
- **Beginner-Friendly**: A good chunk of jobs also accept candidates with no experience.

### Overall Conclusions:

1. **Location Matters**: If you're looking for more job opportunities, bigger cities have more to offer.
2. **Salary Expectations**: The average salaries table can be a useful benchmark for you.
3. **Flexibility in Work**: A fair number of jobs offer remote work and flexible schedules, which could be beneficial if you want a better work-life balance.
4. **Entry-Level Friendly**: The market is welcoming to people with less experience, which is great for someone who's switching fields or just starting.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [369]:
# текст запроса
query_5_1 = '''
select 
    e.name as employer_name,
    count(distinct v.id) as vacancies_cnt

from employers e
left join vacancies v on v.employer_id = e.id
group by employer_name
order by 2 desc
limit 5
'''


In [370]:
# результат запроса
answer_5_1 = pd.read_sql_query(query_5_1, connection)
print('5.1 Топ-5 работодателей по количеству доступных вакансий:')
answer_5_1

5.1 Топ-5 работодателей по количеству доступных вакансий:


employer_name  vacancies_cnt
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [371]:
# текст запроса
query_5_2 = '''
select 
    a.name as area_name,
    count(e.id) as number_of_employers,
    count(v.id) as number_of_job_vacancies
from areas a 
left join employers e on e.area = a.id
left join vacancies v on v.area_id = a.id
where v.id is null
group by area_name
order by number_of_employers desc
limit 1
'''

In [372]:
# результат запроса
answer_5_2 = pd.read_sql_query(query_5_2, connection)
print(f"5.2 Регион с наибольшим числом зарегистрированных работодателей, у которых нет ни одной открытой вакансии:")
answer_5_2

5.2 Регион с наибольшим числом зарегистрированных работодателей, у которых нет ни одной открытой вакансии:


area_name  number_of_employers  number_of_job_vacancies
0    Россия                  410                        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [373]:
# текст запроса
query_5_3 = '''
select
    e.name as employer_name,
    count(distinct v.area_id) as number_of_areas
from employers e
left join vacancies v on v.employer_id = e.id
left join areas a on a.id = v.area_id
group by employer_name
order by 2 desc
limit 10
'''

In [374]:
# результат запроса
answer_5_3 = pd.read_sql_query(query_5_3, connection)
print("5.3 Количество географических областей, в которых у каждого работодателя есть открытые вакансии:")
answer_5_3

5.3 Количество географических областей, в которых у каждого работодателя есть открытые вакансии:


employer_name  number_of_areas
0                      Яндекс              181
1                  Ростелеком              152
2                  Спецремонт              116
3      Поляков Денис Иванович               88
4                   ООО ЕФИН                71
5                  Совкомбанк               63
6                         МТС               55
7  ЭФКО, Управляющая компания               49
8                Почта России               48
9                        КРОН               48

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [375]:
# текст запроса
query_5_4 = '''
select 
    count(*) as cnt
from employers e
left join employers_industries ei on ei.employer_id = e.id
left join industries i on i.id = ei.industry_id
where i.name is null
'''

In [376]:
# результат запроса
answer_5_4 = pd.read_sql_query(query_5_4, connection)
print(f"5.4 Количество работодателей, которые не указали отрасль, в которой они работают: {answer_5_4['cnt'][0]}")

5.4 Количество работодателей, которые не указали отрасль, в которой они работают: 8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [377]:
# текст запроса
query_5_5 = '''
select 
    e.name as employer_name,
    count(distinct i.name) as num_of_industries
from employers e
left join employers_industries ei on ei.employer_id = e.id
left join industries i on i.id = ei.industry_id
group by e.name
having count(distinct i.name) = 4
order by 1
offset 2 limit 1
'''

In [378]:
# результат запроса
answer_5_5 = pd.read_sql_query(query_5_5, connection)
print(f"5.5 Название компании, которое находится на третьем месте в алфавитном порядке среди компаний, работающих в 4 различных отраслях: {answer_5_5['employer_name'][0]}")

5.5 Название компании, которое находится на третьем месте в алфавитном порядке среди компаний, работающих в 4 различных отраслях: 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [379]:
# текст запроса
query_5_6 = '''
select 
    i.name,
    count(distinct e.employer_id) as cnt
from industries i 
join employers_industries e on e.industry_id = i.id
where i.name = 'Разработка программного обеспечения'
group by i.name
'''

In [380]:
# результат запроса
answer_5_6 = pd.read_sql_query(query_5_6, connection)
print(f"5.6 Количество компаний, которые указали 'Разработка программного обеспечения' как отрасль, в которой они работают: {answer_5_6['cnt'][0]}")

5.6 Количество компаний, которые указали 'Разработка программного обеспечения' как отрасль, в которой они работают: 3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [381]:
# код для получения списка городов-милионников
#quote_plus converts the readable url to the string that adheres URL encoding rules
from urllib.parse import quote_plus

#string to convert to URL encoding rules
string_to_encode = 'Города-миллионеры_России'

#converted string to URL encoding standards
encoded_string = quote_plus(string_to_encode)


url = f'https://ru.wikipedia.org/wiki/{encoded_string}'

#read the tables from html
tables = pd.read_html(url)

#extract cities values from Table 1 that contains cities with over 1mil population and convert it to list data type
cities = tables[1]['Город'].values.tolist()

#convert list to tuple type
cities = tuple(cities)


In [382]:
# текст запроса
query_5_7 = f'''
(select 
    a.name as area_name,
    count(v.id) as vacancies_cnt
from employers e 
join vacancies v on v.employer_id = e.id
join areas a on a.id = v.area_id
where a.name in {cities} and e.name = 'Яндекс'
group by area_name

)

union all 

select
    'Total',
    count(v.id)
from employers e 
join vacancies v on v.employer_id = e.id
join areas a on a.id = v.area_id
where a.name in {cities} and e.name = 'Яндекс'

order by 2
'''


In [383]:
# результат запроса
answer_5_7 = pd.read_sql_query(query_5_7, connection)
print(f"5.7 Количество строк в Таблице 5.7 равно: {answer_5_7.shape[0]}")
print("Количество доступных вакансий в компании 'Яндекс' в российских городах с населением более 1 миллиона:")
answer_5_7

5.7 Количество строк в Таблице 5.7 равно: 17
Количество доступных вакансий в компании 'Яндекс' в российских городах с населением более 1 миллиона:


area_name  vacancies_cnt
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

### выводы по анализу работодателей


1. **Доминирующие работодатели**: Компании вроде Яндекса, Ростелекома и Тинькофф являются ключевыми игроками, причём Яндекс значительно опережает других по числу вакансий.
2. **Незаполненные профили работодателей**: Значительное количество работодателей (8419) не указали отрасль, в которой они работают, что может говорить либо о недостатке деталей, либо о разнообразии отраслей, в которых они задействованы.
3. **Географическое присутствие**: У Яндекса и Ростелекома широкий географический охват, с вакансиями в 181 и 152 регионах соответственно. Это отражает их размер и масштаб.
4. **Фокус на отрасли**: С 3553 компаниями, зарегистрированными в отрасли 'Разработка программного обеспечения,' можно сказать, что это высококонкурентная отрасль.

В целом, данные указывают на яркий и разнообразный рынок труда с возможностями в разных регионах и секторах. Однако они также указывают на некоторую концентрацию, с доминирующими игроками вроде Яндекса, занимающими значительную часть рынка.

### English
### Conclusions on Preliminary Data Analysis:

1. **Dominant Employers**: Companies like Yandex, Rostelecom, and Tinkoff are major players, with Yandex far outpacing others in the number of job vacancies.
2. **Unfilled Employer Profiles**: A significant number of employers (8419) have not indicated their industry, indicating either a lack of detail or a diverse set of industries they are involved in.
3. **Geographic Presence**: Yandex and Rostelecom have a broader geographic reach, with vacancies in 181 and 152 areas respectively. This reflects their size and scope.
4. **Industry Focus**: With 3553 companies listed under 'Software Development,' it seems to be a highly competitive industry.

Overall, the data indicates a vibrant and diverse job market with opportunities across different regions and sectors. However, it also points to some concentration, with dominant players like Yandex taking up a significant portion of the market.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [384]:
# текст запроса
query_6_1 = '''
select 
    count(*) as cnt
from vacancies v
where lower(v.name) like '%%data%%' or lower(v.name) like '%%данн%%'
'''

In [385]:
# результат запроса
answer_6_1 = pd.read_sql_query(query_6_1, connection)
print(f"6.1 Количество вакансий, в названии которых содержатся слова 'data' или 'данн': {answer_6_1['cnt'][0]}")

6.1 Количество вакансий, в названии которых содержатся слова 'data' или 'данн': 1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [386]:
# текст запроса
query_6_2 = '''
select
    count(*)
from vacancies v
where (
    v.name ilike '%%data scientist%%' 
    or v.name ilike '%%data science%%'
    or v.name ilike '%%исследователь данных%%'
    or (v.name like '%%ML%%' and v.name not like '%%HTML%%')
    or v.name ilike '%%machine learning%%'
    or v.name ilike '%%машинн%%обучен%%')

and

(
    v.name ilike '%%junior%%'
    or v.experience ilike '%%Нет опыта%%'
    or v.employment ilike '%%Стажировка%%'
)
'''

In [387]:
# результат запроса
answer_6_2 = pd.read_sql_query(query_6_2, connection)
print(f"6.2 Количество доступных вакансий для Data Scientist на начальном уровне без опыта и на волонтерских позициях: {answer_6_2['count'][0]}")


6.2 Количество доступных вакансий для Data Scientist на начальном уровне без опыта и на волонтерских позициях: 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [388]:
# текст запроса
query_6_3 = '''
select
    count(*) as cnt
from vacancies v
where (
    v.name ilike '%%data scientist%%' 
    or v.name ilike '%%data science%%'
    or v.name ilike '%%исследователь данных%%'
    or (v.name like '%%ML%%' and v.name not ilike '%%HTML%%')
    or v.name ilike '%%machine learning%%'
    or v.name ilike '%%машинн%%обучен%%')

and

(
    v.key_skills ilike '%%SQL%%'
    or v.key_skills ilike '%%postgres%%'
)
'''



In [389]:
# результат запроса
answer_6_3 = pd.read_sql_query(query_6_3, connection)
print(f"6.3 Количество вакансий в области Data Science, в которых SQL или Postgres указаны как ключевые навыки: {answer_6_3['cnt'][0]}")

6.3 Количество вакансий в области Data Science, в которых SQL или Postgres указаны как ключевые навыки: 201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [390]:
# текст запроса
query_6_4 = '''
select
    count(*) as cnt
from vacancies v
where (
    v.name ilike '%%data scientist%%' 
    or v.name ilike '%%data science%%'
    or v.name ilike '%%исследователь данных%%'
    or (v.name like '%%ML%%' and v.name not ilike '%%HTML%%')
    or v.name ilike '%%machine learning%%'
    or v.name ilike '%%машинн%%обучен%%')

and

(
    v.key_skills ilike '%%python%%'
)
'''

In [391]:
# результат запроса
answer_6_4 = pd.read_sql_query(query_6_4, connection)
print(f"6.4 Количество вакансий в области Data Science, в которых Python указан как ключевой навык: {answer_6_4['cnt'][0]}")

6.4 Количество вакансий в области Data Science, в которых Python указан как ключевой навык: 351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [392]:
# текст запроса
query_6_5 = '''
select
    round(avg((length(v.key_skills) - length(replace(v.key_skills, CHR(9), '')) + 1)), 2) avg

from vacancies v
where (
    v.name ilike '%%data scientist%%' 
    or v.name ilike '%%data science%%'
    or v.name ilike '%%исследователь данных%%'
    or (v.name like '%%ML%%' and v.name not like '%%HTML%%')
    or v.name ilike '%%machine learning%%'
    or v.name ilike '%%машинн%%обучен%%')
'''

In [393]:
# результат запроса
answer_6_5 = pd.read_sql_query(query_6_5, connection)
print(f"6.5 Среднее количество ключевых навыков для вакансий по Data Science: {answer_6_5['avg'][0]}")

6.5 Среднее количество ключевых навыков для вакансий по Data Science: 6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [394]:
# текст запроса
query_6_6 = '''
select
    v.experience,
    round(avg((coalesce(v.salary_from, v.salary_to) + coalesce(v.salary_to, v.salary_from)) / 2)) avg_salary

from vacancies v
where (
        v.name ilike '%%data scientist%%' 
        or v.name ilike '%%data science%%'
        or v.name ilike '%%исследователь данных%%'
        or (v.name like '%%ML%%' and v.name not like '%%HTML%%')
        or v.name ilike '%%machine learning%%'
        or v.name ilike '%%машинн%%обучен%%'
        )

and   (
        v.salary_from is not null
        or v.salary_to is not null
        )

group by v.experience
'''



In [395]:
# результат запроса
answer_6_6 = pd.read_sql_query(query_6_6, connection)
print(f"6.6 Data Scientist с опытом работы от 3 до 6 лет могут ожидать зарплату в размере: {answer_6_6['avg_salary'][2]} RUB")
answer_6_6

6.6 Data Scientist с опытом работы от 3 до 6 лет могут ожидать зарплату в размере: 243115.0 RUB


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

# выводы по предметному анализу

1. **Популярность Темы**: 1771 вакансии включают в название слова 'data' или 'данн'.
2. **Возможности для Новичков**: 51 вакансия для Data Scientist для начинающих и волонтеров, не требующая опыта.
3. **SQL и Postgres**: 201 вакансия для Data Scientist упоминает SQL или Postgres как ключевые навыки.
4. **Python в Требованиях**: 351 вакансия упоминает Python как ключевой навык.
5. **Среднее Количество Навыков**: В среднем требуется знание 6.41 ключевых навыков для вакансий по Data Science.
6. **Зарплата с Опытом**: Data Scientist с опытом работы от 3 до 6 лет могут ожидать зарплату в размере 243,115 рублей.

В простых словах, в сфере Data Science много вакансий и разные требования к навыкам. Если у тебя есть опыт, особенно с Python или SQL, можно рассчитывать на хорошую зарплату. Даже если ты новичок, есть вакансии для тебя.

### English
### Conclusions on Preliminary Data Analysis in the Field of Data Science:

1. **Topic Popularity**: 1771 vacancies include the words 'data' or 'данн' in the title.
2. **Opportunities for Beginners**: There are 51 job vacancies for Data Scientists at the junior level that require no experience and are volunteer positions.
3. **SQL and Postgres**: 201 Data Scientist vacancies list SQL or Postgres as key skills.
4. **Python Requirements**: 351 vacancies mention Python as a key skill.
5. **Average Number of Skills**: On average, 6.41 key skills are required for Data Science job vacancies.
6. **Salary with Experience**: Data Scientists with 3-6 years of experience can expect a salary of 243,115 RUB.

In simple terms, there are many vacancies in the Data Science field with various skill requirements. If you have experience, especially in Python or SQL, you can expect a good salary. Even if you're a beginner, there are opportunities for you.

# Общий вывод по проекту

In [396]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования


Рынок труда взлетает с почти 50,000 уникальных вакансий от более чем 23,000 работодателей. Возможности географически разнообразны и распределены по различным отраслям, так что у вас есть выбор, оставаться на месте или исследовать новые места и сектора. Большие города, как Москва и Санкт-Петербург, являются ключевыми точками, но меньшие города также набирают обороты. Зарплаты варьируются от 71К до 110К, и есть хороший выбор как полноценных, так и удаленных вакансий. Отличные новости для новичков: многие работы доступны для начинающих.

В области Data Science также много ролей, особенно если вы знаете Python или SQL. Новички не остаются в стороне, и если у вас есть опыт работы от 3 до 6 лет, вы можете рассчитывать на зарплату порядка 243К рублей.

### English
The job market is booming with nearly 50,000 unique vacancies from over 23,000 employers. Opportunities are geographically diverse and spread across different industries, so you've got options whether you want to stay put or explore new places and sectors. Big cities like Moscow and St. Petersburg are hotspots, but smaller ones are catching up. Salaries range around 71K to 110K, and there's a good mix of full-time and remote jobs. Great news if you're a newbie: many jobs are beginner-friendly.

In the Data Science field, there are plenty of roles, especially if you know Python or SQL. Beginners aren't left out either, and if you have 3-6 years of experience, you could earn around 243K RUB.